encoders/firebase-decoders-json/src/json/java/com/google/firebase/decoders/json/JsonDataDecoderBuilderContext.java

```diff
@@ -166,15 +178,42 @@
     ObjectDecoderContextImpl<T> decoderCtx = getObjectDecodersCtx(classToken);
     reader.beginObject();
     while (reader.hasNext()) {
-      String fieldName = reader.nextName();
-      FieldRef<?> fieldRef = decoderCtx.getFieldRef(fieldName);
-      creationCtx.put(fieldRef, decode(fieldRef.getTypeToken()));
+      decodeField(decoderCtx, creationCtx);
     }
     reader.endObject();
     decoderCtx.decodeInlineObjIfAny(creationCtx);
     return creationCtx;
   }
 
+  private <T> void decodeField(ObjectDecoderContextImpl<T> decoderCtx, CreationContextImpl creationCtx) throws IOException {
+    String fieldName = reader.nextName();
+    FieldRef<?> fieldRef = decoderCtx.getFieldRef(fieldName);
+    FieldDescriptor fieldDescriptor = decoderCtx.getFieldDescriptors(fieldName);
+    Object val = decodeField(fieldRef.getTypeToken(), fieldDescriptor);
+    creationCtx.put(fieldRef, val);
+  }
+
+  private <TField, U extends Annotation> TField decodeField(TypeToken<TField> fieldTypeToken, FieldDescriptor fieldDescriptor) throws IOException {
+    TField val = decode(fieldTypeToken);
+
+    //apply annotation modifier
+    for (Class<? extends Annotation> annotationType: fieldDescriptor.getProperties().keySet()) {
+      val = applyFieldModifier(annotationType, fieldDescriptor, val, fieldTypeToken.getRawType());
+    }
+    return val;
+  }
+
+  private <TField, U extends Annotation> TField applyFieldModifier(Class<U> annotationType, FieldDescriptor fieldDescriptor, TField value, Class<TField> fieldType) {
+    @SuppressWarnings("unchecked") //Safe, because each entry in fieldModifier always have the same type parameter
+    FieldModifier<U> modifier = (FieldModifier<U>) fieldModifier.get(annotationType);
+    U annotation = fieldDescriptor.getProperty(annotationType);
+    if (modifier == null)
+      throw new IllegalArgumentException("FieldModifier: " + annotationType + " is not registered.");
```

Comment:
I don't think we should throw in this case. I might not care about a particular annotation but still want to be able to decode.

A hypothetical example: decode pojo from a firebase function call and save into firestore. the pojo could contain `@DocumentId`, but functions don't care about it, so they will just ignore it as opposed to failing to decode.